torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp

```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
+// Find the correct alias representing Value v at Node n.
+Value* InplaceConverter::ValueTracker::findAliasForValueAtNode(
+    Value* v,
+    const Node* n) const {
+  GRAPH_UPDATE("Finding alias for value:", v->debugName(), " at node ", *n);
+  if (alias_to_value_.find(v) == alias_to_value_.end()) {
+    // This value was not affected by any inplace operator.
+    return v;
+  }
+
+  auto* root_v = alias_to_value_.find(v)->second;
+  TORCH_INTERNAL_ASSERT(
+      value_to_sorted_aliases_.find(root_v) != value_to_sorted_aliases_.end());
+  const auto& aliases = value_to_sorted_aliases_.find(root_v)->second;
+
+  // alias is accessible only if
+  // 1. alias owning block is ancestor of n.
+  // 2. alias owning node is before n.
+  // return the last alias that satisfies this condition.
+  Value* found_alias = nullptr;
+  for (auto* alias : aliases) {
+    auto* alias_n = alias->node();
+    if (alias_n->isBefore(n) &&
+        isAncestor(alias_n->owningBlock(), n->owningBlock())) {
+      found_alias = alias;
+    }
+  }
+
+  TORCH_INTERNAL_ASSERT(nullptr != found_alias, "More details: \n",
+    n->sourceRange().str(),
+    "Input ", v->debugName(), " of node ", *n,
+    " was modified by in-place operation, but we cannot find its updated value. ",
+    "Please report a bug to PyTorch, and/or try to avoid using in-place operators on this value.");
+
+  return found_alias;
+}
+
+// Pass over block, and gather the initial value for any attribute.
+// Also cache the full name of the attribute for every GetAttr/SetAttr node.
+void InplaceConverter::gatherAttrNameInitialValueMap(
     Block* block,
-    const std::shared_ptr<Graph>& graph,
-    std::unordered_map<std::string, Value*>& allAttrValues,
-    std::unordered_map<std::string, Value*>& setAttrValues,
-    MutationRemover& mr,
-    Module* module_ = nullptr) {
-  Node* m = *block->nodes().begin();
-  WithInsertPoint guard(m);
-  std::unordered_map<std::string, Value*> nextSetAttrValues = {};
+    std::unordered_map<std::string, Value*>& attr_name_value_map,
+    std::unordered_map<Node*, std::string>& attr_node_fullname_map) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    Node* n = *it;
+    it++; // node n can be destroyed
+
+    for (auto* sub_block : n->blocks()) {
+      gatherAttrNameInitialValueMap(
+          sub_block, attr_name_value_map, attr_node_fullname_map);
+    }
+
+    if (n->kind() != prim::GetAttr && n->kind() != prim::SetAttr)
+      continue;
 
+    auto name = n->s(attr::name);
+    auto attrModule = *module_;
+    Value* paramConst = nullptr;
+
+    auto moduleNames =
+        findSubModuleAttr(n->inputs().at(0), name, attrModule, graph_);
+
+    std::string fullName("");
+    for (auto& name : moduleNames) {
+      fullName += name + '.';
+    }
+    fullName += name;
+
+    attr_node_fullname_map.insert({n, fullName});
+
+    if (attr_name_value_map.find(fullName) == attr_name_value_map.end() &&
```

Comment:
nit: Anyway we can combine this if and L#630 if, in order to avoid trying to find fullname twice

Reply:
L#630 is to check before leaving this function, if `fullname` exist in `attr_name_value_map`, and insert dummy initial value if that's not the case. So the outcome of the two finds is different. But yes there is room to improve the overall logic here.